File: frontend/src/types/history.ts
import type { PromptTransformation } from "@/types/prompt";

export interface PromptHistorySummary {
  id: number;
  weak_prompt: string;
  category: string;
  role: string;
  is_favorite: boolean;
  created_at: string;
}

export interface PromptHistoryRecord
  extends PromptTransformation {
  id: number;
  is_favorite: boolean;
  created_at: string;
  updated_at: string;
}

export interface PromptHistoryListResponse {
  items: PromptHistorySummary[];
  total: number;
  limit: number;
  offset: number;
}

export interface FavoriteUpdateRequest {
  is_favorite: boolean;
}

export interface DeleteHistoryResponse {
  deleted: boolean;
  history_id: number;
}

export interface ClearHistoryResponse {
  deleted_count: number;
}

export interface HistoryQuery {
  search?: string;
  category?: string;
  favoritesOnly?: boolean;
  limit?: number;
  offset?: number;
}

export interface ApiErrorResponse {
  detail?: string;
}

File: backend/src/schemas/history.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from src.schemas.response import PromptTransformation


class PromptHistoryCreate(PromptTransformation):
    """Internal schema used when saving a transformation."""

    pass


class PromptHistoryRead(PromptTransformation):
    """Complete prompt-history record."""

    model_config = ConfigDict(
        from_attributes=True,
    )

    id: int
    is_favorite: bool
    created_at: datetime
    updated_at: datetime


class PromptHistorySummary(BaseModel):
    """Compact history record for lists and sidebars."""

    model_config = ConfigDict(
        from_attributes=True,
    )

    id: int
    weak_prompt: str
    category: str
    role: str
    is_favorite: bool
    created_at: datetime


class PromptHistoryListResponse(BaseModel):
    """Paginated history response."""

    items: list[PromptHistorySummary]
    total: int
    limit: int = Field(ge=1)
    offset: int = Field(ge=0)


class FavoriteUpdateRequest(BaseModel):
    """Request used to update a favorite state."""

    is_favorite: bool


class DeleteHistoryResponse(BaseModel):
    """Response returned after deleting one record."""

    deleted: bool
    history_id: int


class ClearHistoryResponse(BaseModel):
    """Response returned after clearing history."""

    deleted_count: int